### DISCUSSION TITLE
Deepsleep Calculation and Drift

### DISCUSSION THREAD

**Discussion by reimerp** [arendst/Tasmota#13955] [Ideas]
Hi all,

I am running a ESP-M3 battery powered to measure an electricy meter, publishing values to mqtt.
The settings are DeepSleep 3600 / Teleperiod 10
The drift of 4% of the calculated time would result to ~2min for an hour, but since this is +/- the device will wakeup occasionally before the planned time (full hour) and this will confuse clients on mqtt, since sometimes there are two measuments an hours, sometimes none
I did some inverstigation on this: unfortunately smlj was not working as expected, so I ended up adjusting tper in the script to have a second measurement if the script notice that a wakeup happened to early. Basically this is working but not perfect, I now have at least one measurement per hour, but occasinally two just some ninutes between them, also confusing some clients. The device also have to be longer online tthan approx 12s in the early case which drains battery.

Could it be a better approach to calculate DeepSleep to wakeup on the next interval + the max drift, eg not at 12:00:00 but at 12:02:00, than the real wakeup would be in the interval [12:00:00 - 12:04:00] which is always later than the targeted wakeup and never before?

**Comment by barbudor** [comment]
Please share your detailed analysis
My experience with deepsleeptime is that, after a small period of self calibration it is quite accurate within a few seconds of the expect wake up time

**Comment by reimerp** [comment]
I checked this on my live device, and also on an identical hardware (w/o sensor) but identical configuration, upgraded to lastest git commit and also enable mqttlog.
These are sensor timestamps from my live sensor (with active patch -- i added some comments)

<details>

```
2021-12-07T07:00:49Z 2374.455
2021-12-07T08:01:31Z 2374.611   1:31 late   
2021-12-07T08:58:48Z 2374.778   1:12 early
2021-12-07T09:00:48Z 2374.782   corrected by patch
2021-12-07T10:01:23Z 2374.999   1:23 late
2021-12-07T10:58:51Z 2375.124   1:09 early
2021-12-07T11:00:51Z 2375.128   corrected
2021-12-07T12:00:37Z 2375.282   0:37 late
2021-12-07T12:59:40Z 2375.414   0:20 early
2021-12-07T13:00:40Z 2375.416   corrected
2021-12-07T14:01:37Z 2375.556   1:37 late
2021-12-07T14:58:42Z 2375.721   1:18 early
2021-12-07T15:00:42Z 2375.725   corrected
2021-12-07T16:01:21Z 2375.907   1:21 late
```
</details>
I was first thinking of the patch confusing some of the calculations, but it is still present in 10.0.0 w/o any tper adjustments.
I also checked rather old entries were the tper adjustment was not present. The differences to the expected wake up time seems lower but this might be related to battery voltage. There are still readings that were too early and thats why my journey begun.

<details>

```
2021-08-13T16:00:10Z 1831.254
2021-08-13T17:00:20Z 1831.503
2021-08-13T17:59:55Z 1831.718
2021-08-13T19:00:18Z 1831.917
2021-08-13T20:00:11Z 1832.12
2021-08-13T20:59:58Z 1832.304
2021-08-13T22:00:04Z 1832.516
2021-08-13T23:00:08Z 1832.688

```
</details>

**Reply by barbudor** [reply]
From my MQTT broker logs, I looked at the connection time of the sensor and caomputed the delta from the expected wake-up time.
The average on 12K logs is 4 seconds which reasonnably match the time from the wake up, to connect to Wifi and then to MQTT.
91% of the 12K logs are within +/- 3 secs from that
In some rare extend, the error can go close to +/- 30 sec but this is very rare

So yes, there will be jitter, but no drift. When Tasmota finds that there is an error it adjust the duration of the next sleep period.
So for me it is quite good.

Indeed some times Tasmota will be too early, some times a little too late.
I don't see a problem with that.
If you want to present very regular data to your user, apply some quantization to the time. i.e. if you are expected a measurment once per hour, round the time to the closest o'clock, the error is quite small.

Now if you think you can provide a better regulation algorithm, please let us know with the mathematical support that goes with it.

**Reply by sfromis** [reply]
IIRC, the internal timer used by ESP8266 to control when to wake up is not rock stable, meaning that it would take modeling of the factors affecting that timing to avoid jitter in times. It works well enough for most purposes, but some deviation is to be expected. In addition to the jitter from ESP8266, time to connect to wifi can also vary by seconds, or sometimes more.

**Reply by reimerp** [reply]
Yes, for sure there might be deviation in the internal timer, but i think waking up too early (b4 expected time) is worse than too late, cause ist confuses clients that expects values after the time, so the expected wakeup time should reflect this somehow. I can say that the online time is more or less stable around 15s +/- 1s, I tried to optimize:
- used static IPs to avoid DHCP / DNS requests
-  removed unnessesary mqtt requests, like subscribing to group Topic
- avoid logging.
The measurement is the only thing which is retained.
I noticed, that with newer Versions, there is slightly more overhead introduced, by
- INFO messages
- STATUS 1 and STATUS 11 messages (they seems to be issued in the last online cycle)
Can this be avoided by Configuration, just curious, but this doesn't solve the early wakeup.

**Reply by barbudor** [reply]
INFO messages are supposed do be disabled on wake-up restart
STATUS1 and STATUS11 are not supposed to be sent freely. You may have something asking for it (could be a retained message).
Can you please post the logs from start of wake-up to gong back to sleep with Log 3 Level ?

Regarding the time of measurment, if it confuse your users, just round it up to the correct time. You can't avoid a regulation algorithm to oscillate around the average.

**Reply by sfromis** [reply]
Deviation is both ways, you cannot predict which without deep analysis. You may prefer making your interval a bit longer, to avoid too short, but that would of course skew the cycle longer term.

You could also just add 1 minute to any interval you get, and cut down to the expected cycle.... End of confusion

**Reply by reimerp** [reply]
Status1 10 and 11 is issued by homeassistant sorry.

The Deviation Problem.... Maybe i will try to say it in other words: yes Deviation is in both directions, but I just don't want the device to stop sleeping before the next full hour starts. Its a very easy solution:
The sleeptime will still be3600 second, also the correction calculation is untouched, its simply the very first calculation.
Eg. if you issue a deepsleeptime 3600 at 11:23, the first wakeup will be calculated to 12:00, this is because the "too short" branch is triggered. If you just add 5% of deepsleep on top you will get the first wakeup at 12:03. The actual wakeup will be +/- 4% so this might be 12:01 till 12:05 but in every case its later than 12:00. The next calculation simply calculates deviation and adds 3600 seconds again -> 13:03. Voila.

**Reply by reimerp** [reply]
tasmota/xdrv_29_deepsleep.ino

The very first calculation is alwas too short: i think it should be "+" in the if clause

```
+  if (RtcSettings.nextwakeup <= (UtcTime() + DEEPSLEEP_MIN_TIME)) {
+    // ensure nextwakeup is at least in the future, and add 5%
     RtcSettings.nextwakeup += (((UtcTime() + DEEPSLEEP_MIN_TIME - RtcSettings.nextwakeup) / Settings->deepsleep) + 1) * Settings->deepsleep;
+    RtcSettings.nextwakeup += Settings->deepsleep * 0.05;
```
